Modules/_decimal/_decimal.c

```diff
@@ -3241,11 +3243,24 @@ dec_format(PyObject *dec, PyObject *args)
             /* NUL fill character: must be replaced with a valid UTF-8 char
                before calling mpd_parse_fmt_str(). */
             replace_fillchar = 1;
-            fmt = dec_strdup(fmt, size);
-            if (fmt == NULL) {
+            fmt = fmt_copy = dec_strdup(fmt, size);
+            if (fmt_copy == NULL) {
                 return NULL;
             }
-            fmt[0] = '_';
+            fmt_copy[0] = '_';
+        }
+        char *z_start = strchr(fmt, 'z');
+        if (z_start != NULL) {
+            no_neg_0 = 1;
+            size_t z_index = z_start - fmt;
+            if (fmt_copy == NULL) {
+                fmt = fmt_copy = dec_strdup(fmt, size);
```

Comment:
It took me a couple of tries to understand the logic here (in particular, in the case that we have both a `NUL` as a fill character _and_ a `z` in the format string). I'm wondering whether the logic would be clearer if instead of using `fmt_copy` we had a `fmt_is_copy` flag.

Reply:
`fmt_copy` acts as both a pointer to the copy and a flag (NULL if there is no copy)

The original code did use a flag, but I found it sloppy that `fmt` is mutated-- as it originally holds a PyUnicode_AsUTF8AndSize() result that should not be freed.  In the revised code, it's much less likely to have a bug where the PyUnicode_AsUTF8AndSize() result is mistakenly freed.

Reply:
> but I found it sloppy that fmt is mutated

Well, the `fmt` variable is still mutated...  What I meant was that modification to the string itself now always happens through the `fmt_copy` variable.  If such modification happens without having created the copy, it will be a clear runtime error (null ptr access) rather than subtly overwriting the buffer returned by `PyUnicode_AsUTF8AndSize()`.

Reply:
> modification to the copy now always happens through the `fmt_copy` variable

Additionally, I was able to enforce this at compile time by changing `fmt` to `char const *`, so it seems like a win.

Reply:
Okay, fair enough; I guess this comes down to a matter of personal choice. (I personally would rather have the flag separate and explicit rather than implicit.) Resolving here.